---
id: 6846e3e61271e1f164636b13
title: Passo 21
challengeType: 20
dashedName: step-21
---

# --description--

Agora você vai tornar a validação mais granular.Crie uma função chamada `find_invalid_records` para encontrar valores inválidos em um dicionário.Forneça os seguintes parâmetros: `patient_id`, `age`, `gender`, `diagnosis`, `medications` e `last_visit_id`.

Dentro da sua nova função, crie um dicionário vazio chamado `constraints`.Então, retorne `constraints` da sua nova função.

# --hints--

Você deve ter uma função chamada `find_invalid_records`.

```js
({ test: () => runPython(`
assert _Node(_code).has_function("find_invalid_records")
`) })
```

Sua função `find_invalid_records` deve ter os seguintes parâmetros: `patient_id`, `age`, `gender`, `diagnosis`, `medications` e `last_visit_id`, nesta ordem.

```js
({ test: () => runPython(`
assert _Node(_code).find_function("find_invalid_records").has_args("patient_id, age, gender, diagnosis, medications, last_visit_id")
`) })
```

Você deve ter uma variável chamada `constraints` com o valor de um dicionário vazio dentro da sua nova função.

```js
({ test: () => assert(runPython(`
_foo = _Node(_code).find_function("find_invalid_records")
_foo.has_stmt("constraints = {}") or _foo.has_stmt("constraints = dict()")
`)) })
```

Você deve retornar o dicionário `constraints` da sua função `find_invalid_records`.

```js
({ test: () => runPython(`
assert _Node(_code).find_function("find_invalid_records").has_return("constraints")
`) })
```

# --seed--

## --seed-contents--

```py
medical_records = [
    {
        'patient_id': 'P1001',
        'age': 34,
        'gender': 'Female',
        'diagnosis': 'Hypertension',
        'medications': ['Lisinopril'],
        'last_visit_id': 'V2301',
    },
    {
        'patient_id': 'p1002',
        'age': 47,
        'gender': 'male',
        'diagnosis': 'Type 2 Diabetes',
        'medications': ['Metformin', 'Insulin'],
        'last_visit_id': 'v2302',
    },
    {
        'patient_id': 'P1003',
        'age': 29,
        'gender': 'female',
        'diagnosis': 'Asthma',
        'medications': ['Albuterol'],
        'last_visit_id': 'v2303',
    },
    {
        'patient_id': 'p1004',
        'age': 56,
        'gender': 'Male',
        'diagnosis': 'Chronic Back Pain',
        'medications': ['Ibuprofen', 'Physical Therapy'],
        'last_visit_id': 'V2304',
    }
]

--fcc-editable-region--

--fcc-editable-region--

def validate(data):
    is_sequence = isinstance(data, (list, tuple))

    if not is_sequence:
        print('Invalid format: expected a list or tuple.')
        return False
        
    is_invalid = False
    key_set = set(
        ['patient_id', 'age', 'gender', 'diagnosis', 'medications', 'last_visit_id']
    )

    for index, dictionary in enumerate(data):
        if not isinstance(dictionary, dict):
            print(f'Invalid format: expected a dictionary at position {index}.')
            is_invalid = True

        if set(dictionary.keys()) != key_set:
            print(
                f'Invalid format: {dictionary} at position {index} has missing and/or invalid keys.'
            )
            is_invalid = True

    if is_invalid:
        return False
    print('Valid format.')
    return True

validate(medical_records)

```
